compiler/parser/RdrHsSyn.hs

```diff
@@ -2396,6 +2392,48 @@ funBind loc tag mg =
             , fun_tick = []
             }
 
+-- | Compute the free type variables occurring in a choice parameter type.
+--
+-- This is needed to generate choice data types, as we want to drop type variables
+-- which do not occur in the body of the type.
+--
+-- This is a simplistic implementation relying on the fact that we do not support
+-- all types as choice parameters. In particular, we do not support `forall` types
+-- which bind type variables and complicate the calculation.
+--
+-- We throw a parser error if we find an unsupported type.
+freeTypeVarsInChoiceType :: LHsType GhcPs -> P (Set.Set RdrName)
+freeTypeVarsInChoiceType = go
+  where
+    gos :: [LHsType GhcPs] -> P (Set.Set RdrName)
+    gos tys = Set.unions <$> mapM go tys
+    go :: LHsType GhcPs -> P (Set.Set RdrName)
+    go (L loc ty) = case ty of
+      HsForAllTy{} -> addFatalError loc $ text "Forall types not supported in choice argument types"
+      HsQualTy _ _ ty -> go ty
```

Comment:
The context, i.e. the second arg to `HsQualTy` might contain free type variables as well. But we should probably just fail if somebody tries to use `C a => T a` in a choice type.